### PR TITLE
refactor: api v2 schedules

### PR DESCRIPTION
**Pull request by supalarry**
## Problem

v2 `/schedules` endpoints are not intuitive to use e.g. when creating a schedule for Sunday user must pass "0" to represent Sunday, and the response is in shape that atom expects which is messy and hard to understand and contains UI specific data like "isLastSchedule":

```
{
    "status": "success",
    "data": {
        "id": 139,
        "name": "Default",
        "isManaged": false,
        "workingHours": [
            {
                "days": [
                    1,
                    2,
                    3,
                    4,
                    5
                ],
                "startTime": 540,
                "endTime": 1020,
                "userId": 256
            }
        ],
        "schedule": [
            {
                "id": 437,
                "userId": 256,
                "eventTypeId": null,
                "days": [
                    1,
                    2,
                    3,
                    4,
                    5
                ],
                "startTime": "1970-01-01T09:00:00.000Z",
                "endTime": "1970-01-01T17:00:00.000Z",
                "date": null,
                "scheduleId": 139
            }
        ],
        "availability": [
            [],
            [
                {
                    "start": "2024-05-14T09:00:00.000Z",
                    "end": "2024-05-14T17:00:00.000Z"
                }
            ],
            [
                {
                    "start": "2024-05-14T09:00:00.000Z",
                    "end": "2024-05-14T17:00:00.000Z"
                }
            ],
            [
                {
                    "start": "2024-05-14T09:00:00.000Z",
                    "end": "2024-05-14T17:00:00.000Z"
                }
            ],
            [
                {
                    "start": "2024-05-14T09:00:00.000Z",
                    "end": "2024-05-14T17:00:00.000Z"
                }
            ],
            [
                {
                    "start": "2024-05-14T09:00:00.000Z",
                    "end": "2024-05-14T17:00:00.000Z"
                }
            ],
            []
        ],
        "timeZone": "America/Cancun",
        "dateOverrides": [],
        "isDefault": true,
        "isLastSchedule": true,
        "readOnly": false
    }
}
```

## Solution
Simplify create inputs and responses to deliver good developer experience. The API does not distinguish whether the request came from an atom or not - the response is always in the same shape.

On the frontend hooks are used to fetch api data and then transformers to transform api data into atom compatible structure, and then when updating a schedule to transform atom schedule into api compatible structure.

## Endpoints

Whether creating, fetching or updating schedule the response has the same shape.

### POST `/schedules`

Screenshot 2024-05-14 at 12 21 46

### GET `/schedules/default` & GET `/schedules/:id`
Screenshot 2024-05-14 at 12 23 35

### UPDATE `/schedules/id`
Screenshot 2024-05-14 at 12 25 24

### API inputs and outputs
For atom hooks and api to be on the same page, all of schedules inputs e.g. `CreateScheduleInput` and outputs e.g. `GetScheduleOutput` have been moved to platform-types `packages/platform/types/schedules` in order to centralize inputs and outputs of API.

### Data transformation
We want to re-use existing functions while providing a cleaner API interface. That means that:
1. API takes in data defined by the input types, then uses [input-schedules.service.ts](https://github.com/calcom/cal.com/pull/14870/files#diff-0e88b4407fc43de1d8e502f177772e804f10fc4d13b60ac090554be4d530650e) function "transformInputCreateSchedule" that transforms availability days from names e.g "Sunday" to numeric values e.g. 0 aka to data format that is compatible with internal functions. Same is done for overrides. API then passes transformed values to prisma that performs create, read or update operation.
2. We no longer use trpc handlers to update schedule - we completely rely on the [schedules.repository.ts](https://github.com/calcom/cal.com/pull/14870/files#diff-1d22143416e92a38d7ed9ba36d04f19b8e58cd2307fd42b54106591bbb0f11f7) for performing prisma operations.
3. When prisma operation is performed and it returns data, then [output-schedules.service.ts](https://github.com/calcom/cal.com/pull/14870/files#diff-d2d09792efe098b8a004f5e8e3da006a81a606465a20425cd2a96cac676f3a0d) service is used to transform internal data into the "clean" way for response.

If you check "transformInputCreateSchedule" in point 1., then you see that to transform avaiability and overrides it uses functions imported from platform libraries. They reside there not in the class, because they are also used client side - when API data is returned by hook, that data also needs to be converted to data structure compatible with our internal functions. That brings us to client:
1. Hooks return data in the same "clean" structure as API.
3. [transformApiScheduleForAtom.ts](https://github.com/calcom/cal.com/pull/14870/files#diff-7c79b7116227f6b1b66236e191e6d17a2f247f357d74914eb01d0989cdcacae8) is used to transform API response into structure compatible with atom.
4. [transformAtomScheduleForApi.ts](https://github.com/calcom/cal.com/pull/14870/files#diff-0c9c7b435e93e42af524dedf70cf201610577e0c724775da7c726a3e3300173b) is used to transform atom schedule into API structure, for example, when updating schedule.

### New hook - useSchedules
[useSchedules.ts](https://github.com/calcom/cal.com/pull/14870/files#diff-4f601660f9722df81042167369673f749dea1b5b3a07d18a86d8d1cf1146ca95) has been added to be used in [PlatformAvailabilitySettingsWrapper.tsx](https://github.com/calcom/cal.com/pull/14870/files#diff-6ba860894eeb6436c20e300720a396ce07ab7cacfb011112ee4dd8df6ca2b374) - atom needs to know if its users last schedule do disable deleting it. We use `useSchedules` to fetch user schedules, and then pass their count to `transformApiScheduleForAtom`.

## Tests
1. API e2e tests - [schedules.controller.e2e-spec.ts](https://github.com/calcom/cal.com/pull/14870/files#diff-e2df2b72b75c9ef96c2f172ec018e5231e6c428cb2d59f5fd198d74d01311983)
2. Setup jest for atoms ([atoms/jest.config.json](https://github.com/calcom/cal.com/pull/14870/files#diff-bdff82a1dd2d9b620287bca09b0cf7c49ca8637610a192251031f087b862135d)) and then add tests - Atom tests testing the transformation of API data into atom compatible data ([transformApiScheduleForAtom.spec.ts](https://github.com/calcom/cal.com/pull/14870/files#diff-a5879f10930355d634c226d255008c6ffa04eafdb359f925c139052df58b2ebf)) and atom data into API compatible data ([transformAtomScheduleForApi.spec.ts](https://github.com/calcom/cal.com/pull/14870/files#diff-1309c05a304b9fed4f8b8dfa6d25ddf38d0f9871857f5dcc9a8e912241d8ec87))
3. Built atoms and then started v2 api, and in examples app confirm that fetched default schedule is displayed correctly and that its possible to update it too.

## Docs
Swagger docs been updated with examples

# schedules refactor: before & after

### 1. Before

1. availabilities days are numbers instead of names
2. start and end times must be dates even though api only cares about hours and minutes
3. no way to pass overrides - dates and times when availabilities change

Request body:

```jsx
{
  "name": "working hours",
  "timeZone": "Europe/Rome",
  "availabilities": [
    {
      "days": [
        1,
        2,
        3,
        4,
        5
      ],
      "startTime": "1970-01-01T09:00:00.289Z",
      "endTime": "1970-01-01T17:00:00.289Z"
    }
  ],
  "isDefault": true
}
```

Response:

```jsx
{
  "status": "success",
  "data": {
    "id": 256,
    "name": "working hours",
    "isManaged": false,
    "workingHours": [
      {
        "days": [
          1,
          2,
          3,
          4,
          5
        ],
        "startTime": 540,
        "endTime": 1020,
        "userId": 480
      }
    ],
    "schedule": [
      {
        "id": 462,
        "userId": 480,
        "eventTypeId": null,
        "days": [
          1,
          2,
          3,
          4,
          5
        ],
        "startTime": "1970-01-01T09:00:00.000Z",
        "endTime": "1970-01-01T17:00:00.000Z",
        "date": null,
        "scheduleId": 256
      }
    ],
    "availability": [
      [],
      [
        {
          "start": "2024-05-15T09:00:00.000Z",
          "end": "2024-05-15T17:00:00.000Z"
        }
      ],
      [
        {
          "start": "2024-05-15T09:00:00.000Z",
          "end": "2024-05-15T17:00:00.000Z"
        }
      ],
      [
        {
          "start": "2024-05-15T09:00:00.000Z",
          "end": "2024-05-15T17:00:00.000Z"
        }
      ],
      [
        {
          "start": "2024-05-15T09:00:00.000Z",
          "end": "2024-05-15T17:00:00.000Z"
        }
      ],
      [
        {
          "start": "2024-05-15T09:00:00.000Z",
          "end": "2024-05-15T17:00:00.000Z"
        }
      ],
      []
    ],
    "timeZone": "Europe/Rome",
    "dateOverrides": [],
    "isDefault": true,
    "isLastSchedule": true,
    "readOnly": false
  }
}
```

### 2. After

1. `availabilities` body key has been renamed to `availability`
2. Days are words e.g. `Sunday` instead of number e.g. `0`
3. Overrides can be passed.

Screenshot 2024-05-14 at 12 21 46

### 1. Before

1. To update availabilities they had to be passed under `body.schedule` in data format different than when creating a schedule.
2. The response did not include schedule availabilities

Request body:

```jsx
{
  "schedule": [
      [],
      [{ "start": "2022-01-01T10:00:00.000Z", "end": "2022-01-02T16:00:00.000Z" }],
      [],
      [],
      [],
      [],
      []
    ]
}
```

Response:

```jsx
{
  "status": "success",
  "data": {
    "schedule": {
      "userId": 480,
      "name": "working hours",
      "id": 256
    },
    "isDefault": true
  }
}
```

### 2. After

1. Shape is exactly the same as when creating schedule
2. Response includes availabilities and is in same shape as when creating a schedule

Screenshot 2024-05-14 at 12 25 24

### 1. Before
Response is in same shape as when creating.

### 2. After
Response is in same shape as wen creating and updating.
Screenshot 2024-05-14 at 12 23 35

### CODE REVIEW

File: apps/api/v2/src/ee/schedules/schedules.repository.ts [hunk 1]
```diff
@@ -64,6 +96,81 @@ export class SchedulesRepository {
...
+        date: null,
+      });
+    }
+
+    if (overrides) {
+      deleteConditions.push({
+        scheduleId: { equals: scheduleId },
+        NOT: { date: null },
+      });
+    }
```
Comment (ThyMinimalDev) [inline apps/api/v2/src/ee/schedules/schedules.repository.ts:126]: I think this part with delete conditions can use comments to give some context

File: packages/platform/atoms/availability/wrappers/PlatformAvailabilitySettingsWrapper.tsx [hunk 2]
```diff
@@ -33,10 +35,10 @@ export const PlatformAvailabilitySettingsWrapper = ({
   onUpdateError,
   onUpdateSuccess,
 }: PlatformAvailabilitySettingsWrapperProps) => {
-  const { isLoading, data: schedule } = useClientSchedule(id);
-  const mySchedule = schedule as ApiSuccessResponse<ScheduleWithAvailabilitiesForWeb>;
+  const { isLoading, data: schedule } = useSchedule(id);
+  const { data: schedules } = useSchedules();
```
Comment (ThyMinimalDev) [inline packages/platform/atoms/availability/wrappers/PlatformAvailabilitySettingsWrapper.tsx:39]: this seems like a waste to fetch twice schedules, we could just fetch from useSchedules and extract the right schedule from it based on it

File: packages/platform/atoms/vite.config.ts [hunk 3]
```diff
@@ -28,7 +28,7 @@ export default defineConfig({
       fileName: "cal-atoms",
     },
     rollupOptions: {
-      external: ["react", "fs", "path", "os", "react-dom"],
+      external: ["react", "fs", "path", "os", "react-dom", "crypto"],
```
Comment (ThyMinimalDev) [inline packages/platform/atoms/vite.config.ts:31]: do we still need to exclude crypto if atoms is not importing libraries ?

**Comment by supalarry** [comment]
# schedules refactor: before & after

## Creating schedule

### 1. Before

1. availabilities days are numbers instead of names
2. start and end times must be dates even though api only cares about hours and minutes
3. no way to pass overrides - dates and times when availabilities change

Request body:

```jsx
{
  "name": "working hours",
  "timeZone": "Europe/Rome",
  "availabilities": [
    {
      "days": [
        1,
        2,
        3,
        4,
        5
      ],
      "startTime": "1970-01-01T09:00:00.289Z",
      "endTime": "1970-01-01T17:00:00.289Z"
    }
  ],
  "isDefault": true
}
```

Response:

```jsx
{
  "status": "success",
  "data": {
    "id": 256,
    "name": "working hours",
    "isManaged": false,
    "workingHours": [
      {
        "days": [
          1,
          2,
          3,
          4,
          5
        ],
        "startTime": 540,
        "endTime": 1020,
        "userId": 480
      }
    ],
    "schedule": [
      {
        "id": 462,
        "userId": 480,
        "eventTypeId": null,
        "days": [
          1,
          2,
          3,
          4,
          5
        ],
        "startTime": "1970-01-01T09:00:00.000Z",
        "endTime": "1970-01-01T17:00:00.000Z",
        "date": null,
        "scheduleId": 256
      }
    ],
    "availability": [
      [],
      [
        {
          "start": "2024-05-15T09:00:00.000Z",
          "end": "2024-05-15T17:00:00.000Z"
        }
      ],
      [
        {
          "start": "2024-05-15T09:00:00.000Z",
          "end": "2024-05-15T17:00:00.000Z"
        }
      ],
      [
        {
          "start": "2024-05-15T09:00:00.000Z",
          "end": "2024-05-15T17:00:00.000Z"
        }
      ],
      [
        {
          "start": "2024-05-15T09:00:00.000Z",
          "end": "2024-05-15T17:00:00.000Z"
        }
      ],
      [
        {
          "start": "2024-05-15T09:00:00.000Z",
          "end": "2024-05-15T17:00:00.000Z"
        }
      ],
      []
    ],
    "timeZone": "Europe/Rome",
    "dateOverrides": [],
    "isDefault": true,
    "isLastSchedule": true,
    "readOnly": false
  }
}
```

### 2. After

1. `availabilities` body key has been renamed to `availability`
2. Days are words e.g. `Sunday` instead of number e.g. `0`
3. Overrides can be passed.

Screenshot 2024-05-14 at 12 21 46

## Updating schedule

### 1. Before

1. To update availabilities they had to be passed under `body.schedule` in data format different than when creating a schedule.
2. The response did not include schedule availabilities

Request body:

```jsx
{
  "schedule": [
      [],
      [{ "start": "2022-01-01T10:00:00.000Z", "end": "2022-01-02T16:00:00.000Z" }],
      [],
      [],
      [],
      [],
      []
    ]
}
```

Response:

```jsx
{
  "status": "success",
  "data": {
    "schedule": {
      "userId": 480,
      "name": "working hours",
      "id": 256
    },
    "isDefault": true
  }
}
```

### 2. After

1. Shape is exactly the same as when creating schedule
2. Response includes availabilities and is in same shape as when creating a schedule

Screenshot 2024-05-14 at 12 25 24

## Fetching schedule

### 1. Before
Response is in same shape as when creating.

### 2. After
Response is in same shape as wen creating and updating.
Screenshot 2024-05-14 at 12 23 35

File: packages/platform/atoms/vite.config.ts (hunk 3 above)
Comment (supalarry) [reply packages/platform/atoms/vite.config.ts]: Good point - removed `crypto` and still can build atoms. Removed `crypto` in [refactor: remove crypto from atoms vite rollupOptions](https://github.com/calcom/cal.com/pull/14870/commits/c563676d9be4ed5c6b4d5e14cc1d8146bd47f40a)

File: apps/api/v2/src/ee/schedules/schedules.repository.ts (hunk 1 above)
Comment (supalarry) [reply apps/api/v2/src/ee/schedules/schedules.repository.ts]: Done in [refactor: comment schedules repository update logic](https://github.com/calcom/cal.com/pull/14870/commits/9e8606ba3f96678d44ad6f7fcf8392fd222ae72e)

**Review by ThyMinimalDev — APPROVED** [review]

**Pull request merged**